### PR TITLE
Build AppImage

### LINKED ISSUE #8019: AppImage support
**Version of OpenTTD**: 1.9.3
**Expected result**: Ability to drop a single binary for friends/coworkers in UNIX environment to be able to play together in ad-hoc LAN games without obligatory ceremony of installing/building.
**Actual result**: Right now you have to use your package manager in OS which pulls plenty of dependencies, or even build the source if your PM provides old or even no version of OpenTTD. This might increase the latency between idea and playback or even create a serious annoyance for Windows/Mac users yelling at UNIX environment or userbase.

## Possible Solution

Use [AppImage](http://appimage.org). It's a [specification](https://docs.appimage.org/reference/architecture.html) (not a "package manager" or "daemon" or anything) which allows to build standalone, system-independent executables that run on every modern Linux system.

Compared to snap/flatpak it doesn't enforce any 3rd-party tooling on destination machine, and can be considered like Mac DMGs or Windows ("portable") EXEs.

Under the hood it's just a SquashFS loopback filesystem with specified contents wrapped into ELF binary with actual code on top which mounts the FS as unprivileged user and spawn the application in it. It doesn't provide sandboxing (can be done with firejail on demand) in terms of security, just bundles everything what game/application needs in a single binary, thus being quite ideal for OSS games or "larger" self-contained applications like LibreOffice, Krita, and so on. Definitely not a thing for CLI apps, system tools and utiltiies (but no one prevents anyone from building these and they'll work just fine)

## Steps to reproduce

At first I wanted to create a Pull Request for that, adding AppImage as a "bundle" (in `Makefile.bundle.in`) but I found the OpenTTD's build system quite complicated (maybe overcomplicated with that infamous GNU Autotools orchestra). So even if I can rip it out and inject my own inventions here and there, it would be a bit rude and unrespectful to do that.

Instead, I'd like to share the steps which I manually did to achieve working AppImage add my thoughts on it and how OpenTTD behaves:

1. I set up a "most common denominator" OS environment at the time, which I think is Ubuntu 18.04.2 LTS. You can build AppImages on whatever OS you have, but doing so on oldest one you have is recommended.
1.1. I've done that in Docker (Podman) container, but bare-metal, CI, chroot or VM will do.
1.2. For reference, you can recreate the same environment in Docker using the following:
```sh
# "bionic" is codename for Ubuntu 18.04 LTS
docker run -v "$OPENTTD_SOURCES_DIR:/openttd" -ti ubuntu:bionic bash --login
```
2. Got OpenTTD sources from Git in `/openttd` dir, checked out to 1.9.3 tag
3. Installed the following Ubuntu packages resembling the dependencies needed for game:
```
build-essential libsdl1.2-dev zlib1g-dev liblzo2-dev libfluidsynth-dev libfreetype2-dev libicu-dev libfreetype6-dev libfontconfig1-dev libxdg-basedir-dev
```
4. Configured the Autotools in the following way:
```
./configure --prefix-dir=/usr --binary-dir=bin 
```
5. Did `make -j8` to build the game.
6. After I built the game, I needed it installed into the folder which can be directly used by AppImage's AppDir:
```
make install INSTALL_DIR="appdir"
```
7. It's fine for now. We didn't needed any sort of source patching and no change in build systems. Time to build AppImage itself.
8. The AppImage can be build manually, but there's a reference tool called [LinuxDeploy](https://github.com/linuxdeploy/linuxdeploy) which can be used to build AppImages without deep integration into software.
8.1. The release binary is an AppImage itself, so you can `curl` or `wget` it in CI pipeline and use it right there
8.2. The only thing `linuxdeploy` does in the actual `openttd` binary is `RPATH` to make it look for shared libraries in AppDir first, which is the probably the only "trick" making the AppImages work. Dead simple and quite elegant.
8.3. I call `linuxdeploy` like this:
```sh
# "VERSION" is needed to indicate a version of software we build, in other cases it'll use a git/hg/cvs hash and store it in target filename and metadata (optional)
VERSION=1.9.3 ./bin/linuxdeploy--appdir appdir/ --output appimage
```
9. `linuxdeploy` simply calls `ldd` on `openttd`, copies required `*.so`s for it into AppDir (with some exclusions, more on it later), finds `*.desktop` file in AppDir to lookup metadata and icon, packs that into SquashFS+payload and… its done!
9.1. In CIs without `fuse`, you can run `linuxdeploy` (and other AppImages) via `./linuxdeploy --appimage-extract-and-run`
10. We can take a look at it:
```
$ ls *AppImage
OpenTTD-1.9.3-x86_64.AppImage
$  file OpenTTD-1.9.3-x86_64.AppImage
OpenTTD-1.9.3-x86_64.AppImage: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/l, for GNU/Linux 2.6.18, stripped
$ ldd OpenTTD-1.9.3-x86_64.AppImage
        not a dynamic executable
```

### TL;DR version

```sh
docker run -v $HOME/src/openttd:/openttd -ti ubuntu:bionic bash --login
cd /openttd
apt install # ... dependencies ... #
./configure --prefix-dir=/usr --binary-dir=bin 
make -j8
make install INSTALL_DIR="appdir"
wget # ... linuxdeploy URL #
VERSION=1.9.3 ./linuxdeploy --appdir appdir --output appimage
file  OpenTTD-1.9.3-x86_64.AppImage
```

## Issues I found

These are mostly cosmetic or formal ones, which can probably be addressed more officialy by someone not being dumb as me right now:

* **By default, `openttd` will fail with error about missing languages. `strace` shows it also misses all data in from its basedir.**
  * It's quite understandable, as `openttd` will look up its data in `$PWD` or global directory (`/usr/share`) or local user dirs. AppImage **does not fake or remap directories in running process** and `openttd` can't figure out it has all its data stored in `/tmp/.mount_<appimage_name>_<random uuid>/usr/share/games/openttd`.
  * On the other hand, AppImage provides the "$APPDIR" environment variable for all child processes, which stores the absolute AppDir path. That's nice, but we don't want patch the game exclusively for AppImages, right?
  * By default, `linuxdeploy` symlinked `openttd` as `AppRun` which is the default entrypoint for AppImage. It's nice, but it don't need to be like that, the AppRun can be a script to so we don't need to fake binary paths, just use this as `AppRun`:
```sh
#!/usr/bin/env bash
cd "$APPDIR/usr/share/games/openttd" # go to basedir 
"$APPDIR/usr/bin/openttd" $* # "$*" is to pass argument list from AppImage binary call
```
* **I had to bundle OpenGFX/OpenMSX/OpenSFX inside the AppImage (in `$APPDIR/usr/share/games/openttd`)**
  * This is my major pain point of the whole solution, and I can't figure it out.
  * The OpenTTD launches fine without it and asks user for BaNaNaS download, but after I click on "yes, download" button it only shows the progressbar for a fraction of second and it disappears, with completely gray game window.
  * I'm not sure if that's related to AppImage or OpenTTD or my Internet connection, but I'm quite sure the AppImage does not limit/sandbox/prevent filesystems for running binaries
    * As I said it's possible as an option with `firejail`, but should be invoked by user explicitly
    * However, you can make it more "portable", and let AppImage binaries not to pollute the host dirs
      * If there's a folder called `<appimage filename without extension>.home` in your $PWD or AppImage folder, it's being used as a `$HOME` environment variable. Simple and elegant solution.
      * Same goes for `<appimage filename>.config`, which is used as `$XDG_CONFIG_HOME` and that's why I used `libxdg-basedir` in build system
  * Probably that's easy fix or I've just accidently exposed a bug in BaNaNaS client downloader, but on the other side I would like to have a "full standalone" binary with assets combined to maximum convenience for external users, **as long as it remains compatible with OpenTTD assets' license**

* **I had issues with ICU libraries ABI change**:
  * The `linuxdeploy` excludes the `libicu*.so` from integration into AppDir by default, explaining that's a "too internal" system library and it's not often directly called
  * The OpenTTD build system has a `--static-icu` option which I tried to use, but Ubuntu LTS doesn't have `libsicu*.a` libs available. It has, however, `libicu*.a` static libs in its `libicu-dev` packages, so I assume it's just matter of linker configuration and some Autotools magic
  * On the other hand, you can explicitly tell `linuxdeploy` to bundle these libs with you, and resulting AppImage **has been proven working for us on a small network game this evening**. For reference, you can call it like that:
```sh
# you don't need to address .so.XX.y directly, linuxdeploy is a little smarter about symlinks and ABIs
./linuxdeploy \
  --appdir appimage/ \
  -output appimage  \
  --library /usr/lib/x86_64-linux-gnu/libiculx.so  \
  --library /usr/lib/x86_64-linux-gnu/libicutu.so  \
  --library /usr/lib/x86_64-linux-gnu/libicuuc.so  \
  --library /usr/lib/x86_64-linux-gnu/libicuio.so  \
  --library /usr/lib/x86_64-linux-gnu/libicui18n.so  \
  --library /usr/lib/x86_64-linux-gnu/libicudata.so  \
  --library /usr/lib/x86_64-linux-gnu/libicu-le-hb.so
```

* **MIDI music in OpenTTD is non-portable**
  * This is more a development / architectural problem than distribution, but it directly affets this issue.
  * **Both FluidSynth and Timidity++ require a daemon running on host to play music** which is quite problematic for AppImage and its stateless nature.
  * I would like to suggest some library-based software MIDI synthesizer, maybe [WildMIDI](https://github.com/Mindwerks/wildmidi) or anything like that.
  * You can surely play without music, but **I find it lovely and even bringing some nostalgia, as I play OpenTTD for around 15 years**.

## Why I'm doing this here, instead of maintaining it on my own?

**[AppImages are meant to be done by upstream](https://docs.appimage.org/introduction/upstream.html)**. Why?
  * **You can sign it to assume it's going from legitimate source** and make users trust it. It's very optional, but welcoming for users and auxiliary tools can pick that up
  * **You can provide update information**. The AppImage can't update itself, but it can store HTTP/Github Releases/ZSync/FTP URL in metadata to be used in the following ways:
    * By auxiliary tooling like `appimaged`, which is optional but handy in a fleet of AppImages
    * By script in AppRun
    * By application itself in `libappimage` library if you're so kind to support that. The library can also determine if you're running in AppImage or not, so you can use other search paths for baseset.
  * **You can put it in your CI pipelines, show the link on your website and call it a day**: For 99% of the time, they act completely the same as regular Linux binaries, they're just called from lesser-common location and use bundled shared libs.

## A little disclaimer

If you find this like forcing or advertising a new technology, it's not. I'm not affiliated in @probonopd or AppImage project in any way. **The whole reasoning is that I wanted to quickly share an OpenTTD build for many different platforms and make it as most easy as possible, without polluting the host system**. It worked pretty well and I've been looking forward to integrate such distribution form to other users.

P.S.: Dedicated server build can be done that way too, making them deployable in seconds.
P.P.S: Just stumbled upon the fact that your "cousin project" OpenRCT2, **[already does this](https://github.com/OpenRCT2/OpenRCT2/blob/develop/scripts/build-appimage)** and [shares the AppImage build on their downloads page](http://openrct2.org/downloads/releases/latest) which is cool.

### PR DESCRIPTION
**Pull request by azubieta**
This PR adds a Github workflow to pack OpenTDD using the AppImage format from commits and PR to master.
The resulting binary is published as an artifact. 

appimage-builder is ran twice in order to patch the timidity configuration to use freepaths. This could be fixed once the feature gets added into appimage-builder https://github.com/AppImageCrafters/appimage-builder/issues/50

Closes #8019

### CODE REVIEW

File: .github/workflows/build-appimage.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# This is a basic workflow to help you get started with Actions
```
Comment (FLHerne) [inline .github/workflows/build-appimage.yml]: Not true any more!

File: .github/workflows/build-appimage.yml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - name: Checkout
+        uses: actions/checkout@v2
+      
+      - name: Build
+        run: |
+          mkdir -p build && cd build
+          cmake .. -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=Release
+          make -j`nproc`
+
```
Comment (FLHerne) [inline .github/workflows/build-appimage.yml]: Could the existing Ubuntu 20.04 build in `ci-build.yml` produce this as an artifact?

The CPU time might be free to us, but it seems wasteful to run the same compile twice.

Comment (azubieta) [reply .github/workflows/build-appimage.yml]: Yes. I took that workflow as template. The only requirement is to use `/usr` as install prefix. The build type is optional, so you could also have Debug bundles.

File: .github/workflows/build-appimage.yml (hunk 1 above)
Comment (azubieta) [reply .github/workflows/build-appimage.yml]: Ups, will remove it.

**Comment by TrueBrain** [comment]
As IRC doesn't stick, let me write it down here too:

- I wouldn't publish an Artifacts on pull-request; this allows for people to mislead others in downloading binaries that could do not-so-nice things. Currently what is implemented with Azure DevOps that you can say something to a bot, which starts to produce the artifacts. It is so rarely used, I cannot even remember what the command is :P But this is also the reason the CI currently does not attach binaries to Pull Requests. Not because we cannot do that, but because it is unsafe unless marked otheriwse.
- I would rename "appimage.yml" to "appimage-builder.yml" or what-ever the build-tool is called, just to make it clear there are several definition formats for different tools in the appimage eco-system. Otherwise, and I have the IRC logs to proof it, this is rather confusing (especially as the YAML file has no reference to the tool it is meant to be used by; possibly a nice addition too, just to mention: `# Definition file for <url>`, or what-ever floats your boat).
- I would extend the appimage-builder first to allow for scripting of some sort, so the things in the workflow can go in the definition. This allows for people to test out AppImages without reading in the GitHub workflow how to do this. As I would personally put the downloading of the opengfx / opensfx / openmsx in there, as the fixing up of that one file. It would mean people can do: `cmake .. && make && <command for appimage-builder> ../os/linux/appimage-builder.yml`.

As a side-note: OpenTTD currently releases nightlies via Azure DevOps. It is planned to migrate to GitHub Actions, but as long as that migration is not finished, adding AppImages here will not make them appear on the nightlies download pages.

**Comment by FLHerne** [comment]
> I would extend the appimage-builder first to allow for scripting of some sort, so the things in the workflow can go in the definition. This allows for people to test out AppImages without reading in the GitHub workflow how to do this. As I would personally put the downloading of the opengfx / opensfx / openmsx in there, as the fixing up of that one file. It would mean people can do: cmake .. && make && <command for appimage-builder> ../os/linux/appimage-builder.yml.

Alternatively, perhaps add the AppImage build as a CMake target performing these steps, so that users (and the workflow) can just do `cmake .. && make appimage`.

**Comment by azubieta** [comment]
I have renamed the recipe file to `appimage-builder.yml` and added a first line comment linking to the tool documentation. Thanks for the hints. I'll update all my recipes right away.

Regarding the integration with cmake. I would like to improve the tool a bit more before investing time on this.

**Comment by azubieta** [comment]
I've improved the recipe (and appimage-builder), now appimage-builder it's capable of patching stuff and the whole pack process takes a single instruction.

File: os/linux/appimage-builder.yml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      - dpkg-dev
+
+  after_bundle: |
+    echo "source /etc/timidity/freepats.cfg" | tee AppDir/etc/timidity/timidity.cfg
+
+  runtime:
+    path_mappings:
+      - /usr/share/games/openttd:$APPDIR/usr/share/games/openttd
+      - /etc/timidity:$APPDIR/etc/timidity
+      - /usr/lib/timidity:$APPDIR:/usr/lib/timidity
```
Comment (glx22) [inline os/linux/appimage-builder.yml]: Feels wrong when looking the lines above

**Review by glx22 — COMMENTED** [review]
Nice improvement compared to previous versions

File: os/linux/appimage-builder.yml (hunk 3 above)
Comment (azubieta) [reply os/linux/appimage-builder.yml]: Timidity  and its configuration files should to be installed in order to have music. From my point of view the game is not complete without the music :).  But it's up to you whether to include it or not.

Comment (glx22) [reply os/linux/appimage-builder.yml]: I mean the `$APPDIR:/` when all other have `$APPDIR/`

Comment (azubieta) [reply os/linux/appimage-builder.yml]: You're right. Fixed

File: os/linux/appimage-builder.yml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+script: |
+    which unzip || apt-get update ; apt-get install wget unzip -y
+    mkdir -p AppDir/usr/share/icons/hicolor/scalable/apps
+    cp ../media/openttd.svg AppDir/usr/share/icons/hicolor/scalable/apps
+    wget -qcO opengfx.zip https://cdn.openttd.org/opengfx-releases/0.6.0/opengfx-0.6.0-all.zip
+    unzip opengfx.zip -d AppDir/usr/share/games/openttd/baseset
+    wget -qcO openmsx.zip https://cdn.openttd.org/openmsx-releases/0.3.1/openmsx-0.3.1-all.zip
+    unzip openmsx.zip -d AppDir/usr/share/games/openttd/baseset
+    wget -qcO opensfx.zip https://cdn.openttd.org/opensfx-releases/0.2.3/opensfx-0.2.3-all.zip
+    unzip opensfx.zip -d AppDir/usr/share/games/openttd/baseset
```
Comment (TrueBrain) [inline os/linux/appimage-builder.yml]: OpenTTD normally doesn't embed opengfx/openmsx/opensfx in its releases, as it is a choice people can make. As long as freetype/fontconfig is installed, if a user hasn't downloaded any basesets yet, the bootstrap will take it over. Is there any reason this is needed / required in the AppImage?

File: os/linux/appimage-builder.yml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      - python3
+      - python3.6
+      - python3-minimal
+      - python3.6-minimal
+      - perl
+      - perl-base
+      - pkg-config
+      - patch
+      - make
+      - dpkg-dev
```
Comment (TrueBrain) [inline os/linux/appimage-builder.yml]: Please put these lists in alphabet, tnx :)

File: os/linux/appimage-builder.yml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      image: appimagecrafters/tests-env:centos-7
+      command: ./AppRun
+      use_host_x: true
+    ubuntu:
+      image: appimagecrafters/tests-env:ubuntu-xenial
+      command: ./AppRun
+      use_host_x: true
+
+AppImage:
+  update-information: "gh-releases-zsync|OpenTTD|OpenTTD|latest|*x86_64.AppImage.zsync"
```
Comment (TrueBrain) [inline os/linux/appimage-builder.yml:83]: What is this line? Seems to refer to all kinds of things, but I cannot make head or tails from it.

File: os/linux/appimage-builder.yml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  sign-key: None
+  arch: x86_64
```
Comment (TrueBrain) [inline os/linux/appimage-builder.yml:85]: Is AppImage only x86_64, or can it also do i386/i686?

File: os/linux/appimage-builder.yml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    icon: openttd
+    version: latest
+    exec: usr/bin/openttd
+
+  apt:
+    arch: amd64
+    sources:
+      - sourceline: 'deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ focal main restricted universe multiverse'
+      - sourceline: 'deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ focal-updates main restricted universe multiverse'
+      - sourceline: 'deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ focal-backports main restricted universe multiverse'
```
Comment (TrueBrain) [inline os/linux/appimage-builder.yml]: What is backports good for in this case? I am mostly just wondering why it is there, not saying it is wrong or anything!

**Review by TrueBrain — CHANGES_REQUESTED** [review]
We have been working on moving our release workflow to GitHub Actions, so I was wondering if we could integrate this into it already. Found a few things you could possible address for me!

File: os/linux/appimage-builder.yml (hunk 4 above)
Comment (azubieta) [reply os/linux/appimage-builder.yml]: It felt incomplete to me, but we can safely remove this part.

File: os/linux/appimage-builder.yml (hunk 6 above)
Comment (azubieta) [reply os/linux/appimage-builder.yml:83]: Is a sort of code used to tell where can be found new releases for this AppImage. There we say that new releases will be fetched from the OpenTTD Github releases page, and the file names should end with 'x86_64.AppImage.zsync'.

ZSync is the protocol used to partially download the AppImage file. 

More details at: https://github.com/AppImage/AppImageSpec/blob/master/draft.md#update-information

File: os/linux/appimage-builder.yml (hunk 7 above)
Comment (azubieta) [reply os/linux/appimage-builder.yml:85]: Yes, we can make x86, i386, arm64 and armhf builds

Comment (TrueBrain) [reply os/linux/appimage-builder.yml:85]: Could you add all 4 targets for now? We will see how it goes :)

File: os/linux/appimage-builder.yml (hunk 8 above)
Comment (azubieta) [reply os/linux/appimage-builder.yml]: In may be useful in case of there is a backport for a packages that is used. But it could be safely remove as there will be no security critical packages there.

File: os/linux/appimage-builder.yml [hunk 9]
```diff
@@ -0,0 +1,87 @@
+# appimage-builder recipe, see https://appimage-builder.readthedocs.io
+version: 1
+
+script: |
+    which unzip || apt-get update ; apt-get install wget unzip -y
```
Comment (TrueBrain) [inline os/linux/appimage-builder.yml]: Guess unzip is no longer needed now :D Hihi :)

File: os/linux/appimage-builder.yml (hunk 7 above)
Comment (azubieta) [reply os/linux/appimage-builder.yml:85]: I'm not familiar yet with the OpenTTD build systems, and we would have to setup a cross-compile environment for that.

I guess that it's work for a second PR

Comment (TrueBrain) [reply os/linux/appimage-builder.yml:85]: Do you have an example for me how that would look like? Of another project or anything?

File: os/linux/appimage-builder.yml (hunk 9 above)
Comment (azubieta) [reply os/linux/appimage-builder.yml]: indeed

File: os/linux/appimage-builder.yml (hunk 7 above)
Comment (azubieta) [reply os/linux/appimage-builder.yml:85]: https://appimage-builder.readthedocs.io/en/latest/examples/bash.html

The complicated part should be cross-compiling OpenTTD, if you already have this then it's simple.
Here is an example on how to build a bash arm64 AppImage in a x86 host system.

File: os/linux/appimage-builder.yml (hunk 9 above)
Comment (azubieta) [reply os/linux/appimage-builder.yml]: fixed

**Comment by TrueBrain** [comment]
With all the work on the release workflow etc done, this can be updated to be integrated with the new release workflow.

The main issue the AppImage has (as well as the other single-package solutions), and which is kinda a blocker for integrating it with our own releases, is that with the inclusion of `freepats`, the binary becomes unreasonable large. Where most binaries are < 10 MiB, this all of a sudden is ~50 MiB last I checked. The main problem I have with that: it is only for music, which most people disable. So that is a lot of wasted bytes :D

We ran into the same issue with Emscripten, which resulted in packaging a version where sound is just completely disabled. That is not a proper solution for AppImage.

I keep wondering if there isn't a way to slim down `freepats`, even if it is opinionated, to reduce the download size. I know very little about MIDI, but how I understand it, `freepats` allows for multiple different synthesizers, where we in fact only need one, really. It would just mean we pick which one that is, instead of allowing to user to pick one. For packages like this and for example emscripten that is totally acceptable to me, but your-miles-may-vary.

This more as a braindump, as I keep seeing this PR, reading up on it, coming to this same conclusion a few times now :D

**Comment by azubieta** [comment]
I the AppImage can use the "freepats" installation from the system. So if it's too big you can remove it and ship that. But being honest I don't care to give away even 30mb for a complete gaming experience (with sounds).

**Comment by TrueBrain** [comment]
Well, I mention `freepats` as I know that from other platforms being an issue, but I was digging through the AppImage, and there are many more things that could be reduced in size:

- Lot of "changelog.gz" files, which sum up to a nice sum of bytes
- Lot of other unneeded files, bits and pieces
- The full ICU data package (which is as big as `freepats`).

So I take it back, the problem is not with `freepats` on its own; it is a bit bigger :) And if I remember similar talks with `snap` and `flatpack`, we had similar issues there.

In general, I don't mind things like AppImage, but I do not see a way that we can distribute it with our current infrastructure; which means it cannot be part of our release workflow. It might be better if these files are distributed by third-party, as that is already the case with `snap` and `flatpack` (and even Docker images, for that matter).

For me what it comes down to: our goals here don't align. We cannot handle such big files for releases, and these images are, possibly by definition, fat.

This needs some more consideration on my end :)

**Comment by azubieta** [comment]
If your issue is the infrastructure, you can use Github Workflows to build the packages and Github Releases to store the final binaries. This is already done by many others.

The idea is that the users don't have to trust or depend on a third party to run your software.

File: os/linux/appimage-builder.yml [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        key_url: 'http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x3b4fe6acc0b21f32'
+
+    include:
+      - freepats
+      - libicu66
+      - liblzma5
+      - liblzo2-2
+      - libpng16-16
+      - libsdl1.2debian
+      - libsdl2-2.0-0
```
Comment (LordAro) [inline os/linux/appimage-builder.yml:33]: having both of these is redundant

File: os/linux/appimage-builder.yml [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+      - libxdg-basedir1
```
Comment (LordAro) [inline os/linux/appimage-builder.yml:34]: can be removed now

File: os/linux/appimage-builder.yml [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+      - timidity
```
Comment (LordAro) [inline os/linux/appimage-builder.yml:35]: I'd probably want to prefer fluidsynth over timidity

**Comment by TrueBrain** [comment]
As promised, I was going to come back to this.

For our Steam release, we had to invent a `linux-generic`, which we have been pushing out as nightly for the last 2 weeks. This is a bundle that contains the executable + `.so` files you are most likely missing. We create this for amd64, based on CentOS7, which means that most modern Linux installs can run it (+/- 10 years or younger). For us, this is a much easier thing to maintain, and is a relative small download size.

So for us, there is no direct reason to have AppImage, other than giving people options. But I think this is better left for 3rd parties to do so. Similar that we do not publish "official" Docker images, flatpaks, snaps, etc. We leave this, for now at least, to other people do run this. We might reconsider this in the future, but for now, we are fine with this. (and if we change this, we most likely will do it for all these targets, so AppImage, Docker, flatpak, ..).

In general, we removed most other ways we published for Linux, and reduced it to: what-ever CPack can produce. We noticed this is much less likely to break, and at least is a lot easier for us to maintain (read: it requires no maintenance from our side). So there is also no more `rpm` etc in our repo. These all moved downstream, and I am really happy people picked that up and are now maintaining them (seriously, if you are reading this and maintain rpm, flatpak, ..., thank you!)

So that leads me to a conclusion: I am going to close this Pull Request for now. Not, as I made clear in earlier comments, because I don't like this, but more because currently OpenTTD is not ready for this. And this is also why it took so long to come to this result ... because we are a bit conflicted about it :D

Sorry about this, and really thank you for your effort here. Who knows in a few months we get back to this :) Cheers!

**Comment by azubieta** [comment]
Sorry to hear that, I'll setup some regular CI build to keep the binaries up to date.
If you like you can reference the following page as a source for getting the AppImages:
https://www.appimagehub.com/p/1425360/
